Asset creation

jhughes982 opened the pull request:
## Works towards #6 and #7

- This PR adds the frontend and backend functionality to create assets and thereby complete the study creation steps.
- A few additional thoughts:
  - Some more conversations might be needed around the final asset form fields, advice appreciated from @finleybacon 
  - I'm not totally convinced of the UX flow here, I think we will need to have another brainstorming session around how and where to display the assets. Presumably we would want to enable users to click into an asset to edit it over time, so we should think about how best to achieve that.
  - The asset locations might need [defining on the backend](https://github.com/ucl-arc-tre/portal/issues/7#issuecomment-3070210423). Although I think that can be done as a separate PR once this initial asset creation has been reviewed. Happy to do it in this PR though if desired.
  - I'd also like some guidance on adding additional RBAC if needed. As far as I could tell, the asset RBAC access is already controlled via the study role e.g. `/studies/{studyId}/*` which gives access to the assets already via the wildcard. But it's possible we might want some more granular access control when retrieving assets so I'm happy to be advised further on that.
  - I'll add some UI tests once this functionality has been signed off.
 

<!--
For example "Resolves #42" if this PR resolves issue 42
-->

### Checklist

- [ ] Documentation has been updated

jhughes982 commented on internal/service/studies/assets.go:
```diff
@@ -0,0 +1,191 @@
+		return &openapi.AssetCreateValidationError{ErrorMessage: "third_party_agreement is required when accessed_by_third_parties is true"}, nil
+	}
+
+	// Check if the Asset title already exists
+	// var count int64
+	// err := s.db.Model(&types.Asset{}).Where("title = ?", assetData.Title).Count(&count).Error
+	// if err != nil {
+	// 	return nil, types.NewErrServerError(fmt.Errorf("failed to check for duplicate asset title: %w", err))
+	// }
+	// if count > 0 {
+	// 	return &openapi.AssetCreateValidationError{ErrorMessage: fmt.Sprintf("an asset with the title [%v] already exists", assetData.Title)}, nil
+	// }
```
I'm guessing duplicate asset titles are ok? If so I'll delete this bit.

t-young31 replied:
I think so

t-young31 commented on web/src/components/assets/AssetCreationForm.tsx:
```diff
@@ -14,30 +14,13 @@ const AlertMessage = dynamic(() => import("uikit-react-public").then((mod) => mo
 
-export default function AssetForm(props: AssetFormProps) {
-  const { onSubmit, isSubmitting = false } = props;
+export default function AssetCreationForm(props: AssetFormProps) {
```
I find this odd that it's not a modal like the study create form

acholyn replied:
I think making them modals makes sense

acholyn commented:
nit: could we get rid of the radios for these?
<img width="262" height="81" alt="image" src="https://github.com/user-attachments/assets/5eeb99a5-4964-4abc-b209-b9983d176dfb" />

acholyn commented on web/src/components/assets/AssetCreationForm.tsx:
```diff
@@ -411,26 +393,29 @@ export default function AssetForm(props: AssetFormProps) {
-            <option value="Destroyed">Destroyed</option>
+            <option value="active">Active</option>
+            <option value="awaiting">Awaiting</option>
+            <option value="destroyed">Destroyed</option>
```
might be beneficial to expand slightly on what awaiting means (what does it mean?)


also is it likely they will create an asset that's destroyed? This could just be an option for the update form

jhughes982 replied:
Any suggestions on the wording? Perhaps @finleybacon has some ideas?

acholyn replied:
maybe something like
`Active: asset is in environment`
`Awaiting: asset awaiting creation in environment`
`Destroyed: asset has been destroyed in environment`

?

still not really sure of the merit of being able to select destroyed on creation tho

t-young31 commented on web/src/components/assets/AssetCreationForm.tsx:
Form thoughts:

 - I think the legal basis question could do with some helper text
 - Do assets have a expiry date?
 - I'm not sure the 3rd party agreements selection is needed? I assume contracts will be added onto and asset after it's been created?

finleybacon replied:
> I think the legal basis question could do with some helper text

https://ico.org.uk/for-organisations/uk-gdpr-guidance-and-resources/lawful-basis/a-guide-to-lawful-basis/
is canon, is it good practice to just throw a link into the form? it's a lot of words. if necessary it could be:
 "You must have a valid lawful basis in order to process personal data. There are six available lawful bases for processing set out in Article 6 of the UK GDPR. At least one of these must apply whenever you process personal data."

> Do all assets have a expiry date?

I think no. Currently Stage 4 of the IG process (the final sign off stage) expires every 12 months, which prompts the IAO to look over all the Study data, incl contracts and assets, and ensure it's all still accurate. I think it would be good to recreate that rather that rather than have individual assets expiring. If the asset has an associated 3rd party contract, then of course the contract can expire, and so would be cool if that shows up in the UI when you're looking at a linked asset?

>I'm not sure the 3rd party agreements selection is needed? I assume contracts will be added onto and asset after it's been created?

Good question. Is it possible for this field to be a dropdown lookup of existing contracts they've created. and if it's they don't want to link it to an existing one then could you have a "link to new contract" option in that creates a new contract record _from the asset form_? but not a required field in any case

jhughes982 replied:
Thanks @finleybacon !

I have added https://github.com/ucl-arc-tre/portal/pull/250/commits/7c76b32cca1d73b800a0e9665e9c6440e079d51b regarding the legal basis helper text. Happy to tweak it further though if needed.

Regarding the final two points (expiry and 3rd part agreements) I will leave those for Monday.

t-young31 commented on web/src/components/assets/Assets.tsx:
```diff
@@ -1,59 +1,153 @@
+        <AssetCreationForm handleAssetSubmit={handleAssetSubmit} />
+      ) : (
+        <div>
+          <div className={styles["assets-summary"]}>
+            <span>Assets for this study:</span>
+            <span className={styles["assets-count-badge"]}>{studyAssets.length}</span>
+          </div>
```
I'm not sure this is that useful(?)

acholyn replied:
i vote keep unless we're told it's not helpful

acholyn commented on web/src/components/assets/AssetCard.module.css:
```diff
@@ -0,0 +1,111 @@
+.asset-card {
+  width: 50%;
+  margin: auto;
```
```suggestion
  margin: auto;
  margin-top: 1rem;
```

t-young31 commented on web/src/components/assets/Assets.tsx:
```diff
@@ -1,59 +1,153 @@
+          </div>
+
+          <div className={styles["asset-actions"]}>
+            <Button onClick={() => setShowAssetForm(!showAssetForm)} variant="secondary">
```
needs some padding 
<img width="712" height="129" alt="Screenshot 2025-08-29 at 13 33 57" src="https://github.com/user-attachments/assets/5a8d3b95-3e4e-42b5-922c-417fcfbd713a" />

jhughes982 replied:
[Done](https://github.com/ucl-arc-tre/portal/pull/250#discussion_r2310047554)

t-young31 commented on web/src/components/assets/Assets.tsx:
```diff
@@ -1,59 +1,153 @@
+
+          <div className={styles["asset-actions"]}>
+            <Button onClick={() => setShowAssetForm(!showAssetForm)} variant="secondary">
+              {showAssetForm ? "Cancel" : "Create Another Asset"}
```
```suggestion
              {showAssetForm ? "Cancel" : "Add Asset"}
```
?

t-young31 commented on web/src/components/assets/Assets.tsx:
```diff
@@ -1,59 +1,153 @@
+
+          <div className={styles["assets-grid"]}>
+            {studyAssets.map((asset) => (
+              <AssetCard key={asset.id} asset={asset} />
```
this card is a different style to the study card – I think they should be the same

jhughes982 replied:
I've aligned the styles more now. Anything else you think needs changing?

t-young31 commented on web/src/components/assets/AssetCard.tsx:
```diff
@@ -0,0 +1,88 @@
+      <div className={styles["asset-header"]}>
+        <h4 className={styles["asset-title"]}>{asset.title}</h4>
+        <div className={styles["asset-meta"]}>
+          <span className={`${styles["asset-badge"]} ${getClassificationClass(asset.classification_impact)}`}>
+            {formatClassification(asset.classification_impact)}
+          </span>
```
might be nice pulled right

jhughes982 replied:
Hmm, not quite sure what you are referring to there. They are pulled to the right of the card on my screen.

<img width="604" height="215" alt="Screenshot 2025-08-29 at 16 46 59" src="https://github.com/user-attachments/assets/5e0d3df8-5ee9-41af-b496-63ce71344da9" />

t-young31 commented on web/src/components/studies/ManageStudy.tsx:
```diff
@@ -11,43 +14,51 @@ type ManageStudyProps = {
+        isComplete={studyStepsCompleted}
         completionTitle="Study Setup Complete!"
         completionSubtitle="You have successfully completed all study setup steps."
         completionButtonText="Go to studies"
```
I'm not sure how I feel about having two back buttons

jhughes982 replied:
Happy to remove. Any thoughts @acholyn @finleybacon ?

jhughes982 replied:
https://github.com/ucl-arc-tre/portal/pull/250/commits/93c16c29097c855a9cfec537a33120d9fbd4575e

acholyn commented on web/src/components/assets/Assets.tsx:
```diff
@@ -1,59 +1,153 @@
+  }, [studyId, setAssetManagementCompleted]);
+
+  const handleAssetSubmit = async (assetData: AssetFormData) => {
+    console.log("Creating asset:", assetData);
```
```suggestion
```

t-young31 commented on internal/handler/web/studies.go:
```diff
@@ -48,6 +49,32 @@ func studyToOpenApiStudy(study types.Study) openapi.Study {
+
+	return openapi.Asset{
+		Id:                     asset.ID.String(),
+		UserId:                 &AssetUserIDStr,
+		StudyId:                &AssetStudyIDStr,
```
these can be required fields(?)

jhughes982 replied:
https://github.com/ucl-arc-tre/portal/pull/250/commits/1543375b77cbc2be972b7e514b09c7cc0a2d8916

t-young31 commented on internal/handler/web/studies.go:
```diff
@@ -116,40 +143,33 @@ func (h *Handler) GetStudiesStudyIdAssets(ctx *gin.Context, studyId string) {
-		"message": "Study asset creation is not yet implemented.",
-		"studyId": studyId,
-	})
+	log.Info().Msg("Asset creation post route")
```
```suggestion
```

t-young31 commented on internal/handler/web/studies.go:
```diff
@@ -116,40 +143,33 @@ func (h *Handler) GetStudiesStudyIdAssets(ctx *gin.Context, studyId string) {
-	})
+	log.Info().Msg("Asset creation post route")
+
+	AssetData := openapi.AssetBase{}
```
`assetData`

t-young31 commented on internal/service/studies/assets.go:
```diff
@@ -0,0 +1,191 @@
+
+func (s *Service) validateAssetData(assetData openapi.AssetBase) (*openapi.AssetCreateValidationError, error) {
+	// Validate title
+	if strings.TrimSpace(assetData.Title) == "" {
```
I'd be tempted to define regex patterns for all of these fields and do away with the `TrimSpace` etc.

t-young31 commented on internal/service/studies/assets.go:
```diff
@@ -0,0 +1,191 @@
+	return nil, nil
+}
+
+func (s *Service) CreateAsset(ctx context.Context, user types.User, assetData openapi.AssetBase, studyID string) (*openapi.AssetCreateValidationError, error) {
```
for consistency I think the studyID should be a UUID

t-young31 commented on internal/service/studies/assets.go:
```diff
@@ -0,0 +1,191 @@
+}
+
+func (s *Service) CreateAsset(ctx context.Context, user types.User, assetData openapi.AssetBase, studyID string) (*openapi.AssetCreateValidationError, error) {
+	log.Info().Msg("Starting asset creation")
```
```suggestion
	log.Debug().Any("studyID", studyID).Any("user", user).Msg("Creating asset")
```

t-young31 commented on internal/types/studies.go:
```diff
@@ -58,6 +58,7 @@ type StudyAdmin struct {
 
 type Asset struct {
 	ModelAuditable
+	UserID                 uuid.UUID `gorm:"not null;index"`
```
I wonder if something like `CreatorUserID` would be better. or is it obvious that `asset.UserID` is the person who created it?

t-young31 reviewed (commented):
lovely @jhughes982

finleybacon commented:
> > is it always clear to the asset owner what the legal basis should be phrased as or would some sort of help text be beneficial?
> 
> Some help text could indeed be beneficial here. @finleybacon any ideas?

suggestion here https://github.com/ucl-arc-tre/portal/pull/250#discussion_r2310323763

jhughes982 commented:
> style updates

Done here: https://github.com/ucl-arc-tre/portal/pull/250/commits/c52b1fcffceba8e51f10b0b81503e7e504fc60c6

jhughes982 commented on web/src/components/assets/AssetCard.module.css:
```diff
@@ -0,0 +1,111 @@
+.asset-card {
+  width: 50%;
```
https://github.com/ucl-arc-tre/portal/pull/250/commits/4b94395327d6c9443b32498ae8c2355589f37c44

I think that works a bit better. Although the site in general hasn't been optimised for very small screens yet.

jhughes982 commented:
I have now addressed all the above comments. I have also added some frontend tests of the study management process. So another round of feedback is appreciated.

t-young31 commented on internal/types/studies.go:
```diff
@@ -58,23 +58,23 @@ type StudyAdmin struct {
+	Protection           string    `gorm:"not null"`
+	LegalBasis           string    `gorm:"not null"`
+	Format               string    `gorm:"not null"`
+	Expiry               string    `gorm:"not null"`
```
I think this should be a `time.Time` and called `ExpiresAt` (will need to be careful about the migration as gorm doesn't drop columns automatically)

acholyn reviewed (approved):
😍

jhughes982 commented on web/cypress/fixtures/assets-with-sample.json:
```diff
@@ -0,0 +1,42 @@
+    "has_dspt": false,
+    "stored_outside_uk_eea": false,
+    "accessed_by_third_parties": true,
+    "third_party_agreement": "test789",
```
https://github.com/ucl-arc-tre/portal/pull/250#:~:text=removed%20unused%20fixture%20attributes

jhughes982 commented on internal/types/studies.go:
```diff
@@ -58,20 +58,18 @@ type StudyAdmin struct {
-	HasDspt                bool      `gorm:"not null;default:false"`
-	StoredOutsideUkEea     bool      `gorm:"not null;default:false"`
-	AccessedByThirdParties bool      `gorm:"not null;default:false"`
-	ThirdPartyAgreement    string    `gorm:""`
```
I have added some migration code. I tested this locally and it does seem to work. But I'd appreciate some more guidance as this is the first time running a custom Gorm migration. 

(I also addressed your earlier comment about the time.Time error return and adding a const)
